dependency-specification.rst

```diff
@@ -0,0 +1,376 @@
+dependencies are only relevant in one platform, or only some versions are
+acceptable, so the language permits describing all these cases.
+
+The language defined is a compact line based format which is already in
+widespread use in pip requirements files, though we do not specify the command
+line option handling that those files permit. There is one caveat - the
+URL reference form, specified in PEP-440 [#pep440]_ is not actually
+implemented in pip, but since PEP-440 is accepted, we use that format rather
+than pip's current native format.
+
+Motivation
+==========
+
+Any specification in the Python packaging ecosystem that needs to consume
+lists of dependencies needs to build on an approved PEP for such, but 
+PEP-426 [#pep426]_ is mostly aspirational - and there are already existing
+implementations of the dependency specification which we can instead adopt.
+The existing implementations are battle proven and user friendly, so adopting
+them is arguably much better than approving an aspirational, unconsumed, format.
+
+Specification
+=============
+
+Examples
+--------
+
+All features of the language shown with a name based lookup::
+
+    requests \
+        [security] >= 2.8.1, == 2.8.* \
+        ; python_version < "2.7.10" \
+        # Fix HTTPS on older Python versions.
+
+A minimal URL based lookup::
+
+    pip @ https://github.com/pypa/pip/archive/1.3.1.zip#sha1=da9234ee9982d4bbb3c72346a6de940a148ea686
+
+Concepts
+--------
+
+A dependency specification always specifies a distribution name. It may
+include extras, which expand the dependencies of the named distribution to
+enable optional features. The version installed can be controlled using
+version limits, or giving the URL to a specific artifact to install. Finally
+the dependency can be made conditional using environment markers.
+
+Grammar
+-------
+
+We first cover the grammar briefly and then drill into the semantics of each
+section later.
+
+A distribution specification is written in ASCII text. We use ABNF [#abnf]_ to
+provide a precise grammar. Specifications may have comments starting with a
+'#' and running to the end of the line::
+
+    comment       = "#" *(WSP / VCHAR)
+
+Specifications may be spread across multiple lines if desired using
+continuations - a single backslash followed by a new line ('\\\\n')::
+
+    CSP           = 1*(WSP / ("\\" LF))
+
+Versions may be specified according to the PEP-440 [#pep440]_ rules. (Note:
+URI is defined in std-66 [#std66]_::
+
+    version-cmp   = "<" / "<=" / "!=" / "==" / ">=" / ">" / "~=" / "==="
+    version       = 1*( DIGIT / ALPHA / "-" / "_" / "." / "*" )
+    versionspec   = ["("] version-cmp version *(',' version-cmp version) [")"]
+    urlspec       = "@" URI
+
+Environment markers allow making a specification only take effect in some
+environments::
+
+    marker-op     = version-cmp / "in" / "not in"
+    python-str-c  = (WSP / ALPHA / DIGIT / "(" / ")" / "." / "{" / "}" /
+                    "-" / "_" / "*"
```

Comment:
I'll a note. I'm being deliberately conservative - none of the vars contain escape-worthy characters today so there's no need for escapes.